crypto/certificate.go

```diff
@@ -0,0 +1,114 @@
+package crypto
+
+import (
+	"errors"
+)
+
+// Signable defines an interface which would generate a unique hash based on its content
+type Signable interface {
+	Hash() []byte
+}
+
+// PublicKey defines an interface of Public Key
+type PublicKey interface {
+	Bytes() []byte
+	Str() string
+}
+
+// Signature defines an interface of
+type Signature interface {
+	Bytes() []byte
+}
+
+// Verifier defines an interface which could verify a signature
+type Verifier interface {
+	Verify(signature Signature) bool
+	PublicKey() PublicKey
+}
+
+// Signer defines an interface which could sign a signable
+type Signer interface {
+	Verifier
+	Sign(Signable) (Signature, error)
+}
+
+// Certificate represents a certificate for a signable
+type Certificate struct {
+	signable   Signable
+	endorsers  []PublicKey
+	signatures map[string]Signature
+}
+
+// NewCertificate returns a certificate draft given a signable object
+func NewCertificate(signable Signable) (*Certificate, error) {
+	if signable == nil {
+		return nil, errors.New("Cannot create a certificate with signable as nil")
+	}
+	return &Certificate{
+		signable:   signable,
+		endorsers:  []PublicKey{},
+		signatures: make(map[string]Signature),
+	}, nil
+}
+
+// HashOfSignable returns the hash of content to ceritificate
+func (c *Certificate) HashOfSignable() []byte {
+	return c.signable.Hash()
+}
+
+// Endorse adds the endorser's signature to the certificate
+func (c *Certificate) Endorse(endorser Signer) error {
+	signature, err := endorser.Sign(c.signable)
+	if err != nil {
+		return err
+	}
+	pk := endorser.PublicKey()
+	if _, ok := c.signatures[pk.Str()]; ok {
+		return errors.New("duplicate endorser")
+	}
+	c.endorsers = append(c.endorsers, pk)
+	c.signatures[pk.Str()] = signature
+
+	return nil
+}
+
+// Endorsers returns the list of endorsers of this certificate
+func (c *Certificate) Endorsers() []PublicKey {
+	return c.endorsers
+}
+
+// Signature returns the signature of the verifier saved in this certificate
+func (c *Certificate) Signature(pk PublicKey) Signature {
+	if pk == nil {
+		return nil
+	}
+	signature, ok := c.signatures[pk.Str()]
+	if !ok {
+		return nil
+	}
+
+	return signature
+}
+
+// VerifiedBy checks whether the certificate contains verifier's signature
+func (c *Certificate) VerifiedBy(verifier Verifier) bool {
+	pkey := verifier.PublicKey()
+	if signature := c.Signature(pkey); signature != nil {
+		return verifier.Verify(signature)
+	}
+
+	return false
+}
+
+// VerifiedByAll checks whether the certificate contains the signatures of all these verifiers
+// and only these verifiers
+func (c *Certificate) VerifiedByAll(verifiers []Verifier) bool {
```

Comment:
can we combine `VerifiedBy` and `VerifiedByAll` to `VerifiedBy(verifiers ...Verifier)`?